[WIP] bazel, tensorflow: enable Darwin source build

lukeadams opened the pull request:
###### Motivation for this change
Tracking for WIP patches.

 * Need to break apart commits since some Bazel commits are mixed in with the tf commit
 * need to set LD to re-enable bazel tests on Darwin
 * also need to make the changes Darwin only since linux doesn't need these hacks. usually just lib.optionalString
###### Things done
With this PR tf is buildable for py27/py36, sidestepping the unicode mismatch when using a wheel for py27. This also enables tf c++ artifacts, but I haven't tested those yet.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
bazel: add JAVA_HOME to wrapProgram

removes need to include jdk/jre in dependent packages

Commit:
bazel: enable Darwin building

hacky method to solidify NIX_* env vars as Bazel does not pass these through when calling cc/cxx

matthewbauer commented on pkgs/development/tools/build-managers/bazel/default.nix:
```diff
@@ -71,7 +92,7 @@ stdenv.mkDerivation rec {
   # Build the CPP and Java examples to verify that Bazel works.
 
-  doCheck = true;
+  doCheck = false;
```
This should be a conditional, I think.

lukeadams replied:
I'll change this to `(!stdenv.isDarwin)` if I can't get tests working

matthewbauer commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -105,11 +109,32 @@ let
+        cp test.sh cxx.sh
+        cp test.sh ld.sh
+
+        echo "(clang \"\$@\" ) || (clang++ \"\$@\" )" >> cc.sh
```
Have you tried just doing "clang++" for this? I think that clang++ should support both (although some of the Nix purity stuff may break this).

lukeadams replied:
I initially tried this, but it caused failures when building a few dependencies. I'll try to find a log so I can at least put a comment explaining it.

matthewbauer commented:
Thanks for working on this! Tensorflow on Darwin would be great.

abbradar commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -88,7 +92,7 @@ let
     hardeningDisable = [ "all" ];
 
-    bazelFlags = [ "--config=opt" ]
+    bazelFlags = [ "--config=opt" "--config=monolithic" ]
```
Please leave a comment on why is this necessary; let's make it optional for Darwin to avoid static dependencies as much as possible.

abbradar commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -58,7 +61,8 @@ let
 
     buildInputs = [ python jemalloc openmpi glibcLocales numpy ]
-      ++ lib.optionals cudaSupport [ cudatoolkit cudnn nvidia_x11 ];
+      ++ lib.optionals cudaSupport [ cudatoolkit cudnn nvidia_x11 ]
```
Mark `stdenv.isDarwin && cudaSupport` as broken.

lukeadams replied:
👍

lukeadams replied:
Would it be a good idea to remove the asserts at the top? (see https://github.com/NixOS/nixpkgs/issues/36229)

LnL7 commented:
@GrahamcOfBorg build bazel

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=e6f56c92-293e-4b20-b4ce-34d0cbda4904)

The following builds were skipped because they don't evaluate on aarch64-linux: bazel

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=25c18b49-57c9-44ae-b4db-687e784ca9bb)

Attempted: bazel

<details><summary>Partial log (click to expand)</summary><p>

```
Target //src:bazel failed to build
[6 / 177] no action

INFO: Elapsed time: 5.008s, Critical Path: 0.07s
[6 / 177] no action

FAILED: Build did NOT complete successfully


ERROR: Could not build Bazel
builder for ‘/nix/store/qi5r9dr1xg5ryz972nkif7hwm7d434qr-bazel-0.10.1.drv’ failed with exit code 1
error: build of ‘/nix/store/qi5r9dr1xg5ryz972nkif7hwm7d434qr-bazel-0.10.1.drv’ failed
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=fa407d85-3925-45d7-8ba4-746e160d13bf)

Attempted: bazel

<details><summary>Partial log (click to expand)</summary><p>

```
    [Sched] Linking external/com_google_protobuf/js_embed [for host]
Target //src:bazel failed to build
[32m[1,179 / 1,914][0m no action
[32mINFO: [0mElapsed time: 622.827s, Critical Path: 173.37s
[32m[1,179 / 1,914][0m no action
[31m[1mFAILED:[0m Build did NOT complete successfully
[0m
ERROR: Could not build Bazel
builder for '/nix/store/nwzhb2hkqacflb2jw23f61c4a51539r8-bazel-0.10.1.drv' failed with exit code 1
[31;1merror:[0m build of '/nix/store/nwzhb2hkqacflb2jw23f61c4a51539r8-bazel-0.10.1.drv' failed
```
</p></details>

Commit:
tensorflow: enable Darwin building

bazel cc/cxx hacks on darwin

lukeadams commented:
@abbradar I cleaned it up quite a bit and added conditionals on everything! There are still a few whitespaces changes I need to make though. Rebased onto master, and am currently testing the py27 build (py36 succeeded).

LnL7 commented:
@GrahamcOfBorg build bazel

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=2a3f0f70-2360-4182-b920-29a529bc346e)

The following builds were skipped because they don't evaluate on aarch64-linux: bazel

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=51a7ad7b-ebae-4245-a38c-b811f047a0f6)

Attempted: bazel

<details><summary>Partial log (click to expand)</summary><p>

```
//examples/java-native/src/test/java/com/example/myproject:hello         PASSED
 in 0.4s

Executed 2 out of 2 tests: 2 tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
installing
post-installation fixup
patching script interpreter paths in /nix/store/6rdvkgs8dk2ypyi962chgsj1mldppvjw-bazel-0.10.1
checking for references to /tmp in /nix/store/6rdvkgs8dk2ypyi962chgsj1mldppvjw-bazel-0.10.1...
/nix/store/6rdvkgs8dk2ypyi962chgsj1mldppvjw-bazel-0.10.1
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=a4307a0b-858c-44ce-b38f-e7778cbb1cbd)

Attempted: bazel

<details><summary>Partial log (click to expand)</summary><p>

```
    [Sched] Linking third_party/zlib/libzlib.a
Target //src:bazel failed to build
[32m[687 / 952][0m no action
[32mINFO: [0mElapsed time: 311.403s, Critical Path: 182.49s
[32m[687 / 952][0m no action
[31m[1mFAILED:[0m Build did NOT complete successfully
[0m
ERROR: Could not build Bazel
builder for '/nix/store/61nlfs2w8xv3r1ahgxx5nmjwi9c9brch-bazel-0.10.1.drv' failed with exit code 1
[31;1merror:[0m build of '/nix/store/61nlfs2w8xv3r1ahgxx5nmjwi9c9brch-bazel-0.10.1.drv' failed
```
</p></details>

LnL7 commented:
Do you have the CLT/Xcode installed?

lukeadams commented:
@LnL7 Ahh yep (just CLT) – good catch. Since I need it on my main machine, I can work it out Friday on a VM without it installed.

Sorry about that!

LnL7 commented:
That might explain the missing framework, if you add it I can start another build.

Commit:
bazel: add foundation.framework

lukeadams commented:
@abbradar I'm not going to have time to finish this for the next month, so maybe you can merge the wheel as python3 only? 

@LnL7 I added the framework, but there are other issues due to the way Bazel calls xcrun/xcbuild. It may be simple to patch, but I don't have time to play with it

LnL7 commented:
@GrahamcOfBorg build bazzel

GrahamcOfBorg commented:
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=519a2d9b-9005-4e4c-af2b-5c714bb85d66)

The following builds were skipped because they don't evaluate on x86_64-linux: bazzel

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `bazzel' because:
error: attribute 'bazzel' in selection path 'bazzel' not found

```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=59fc4e46-7d60-4d31-830e-cced9166c078)

The following builds were skipped because they don't evaluate on aarch64-linux: bazzel

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `bazzel' because:
[31;1merror:[0m attribute 'bazzel' in selection path 'bazzel' not found

```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=565dc0cf-2ffd-40b1-a260-bb9ef856dec6)

The following builds were skipped because they don't evaluate on aarch64-linux: bazel

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=83fee279-35a8-4d2d-8a79-d8b014b18957)

The following builds were skipped because they don't evaluate on x86_64-darwin: bazzel

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `bazzel' because:
[31;1merror:[0m attribute 'bazzel' in selection path 'bazzel' not found

```
</p></details>

lukeadams commented:
@LnL7 The framework was in my staging branch; I just pushed it to this one

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=b559839e-588b-4dc3-b1bd-36a7215c4919)

Attempted: bazel

<details><summary>Partial log (click to expand)</summary><p>

```
    Linking third_party/protobuf/3.4.0/libprotobuf.a [for host]; 0s local
Target //src:bazel failed to build
[32m[270 / 298][0m no action
[32mINFO: [0mElapsed time: 253.583s, Critical Path: 45.79s
[32m[270 / 298][0m no action
[31m[1mFAILED:[0m Build did NOT complete successfully
[0m
ERROR: Could not build Bazel
builder for '/nix/store/nygjcxa5y2n0xascw1n913fjbfz874d7-bazel-0.11.1.drv' failed with exit code 1
[31;1merror:[0m build of '/nix/store/nygjcxa5y2n0xascw1n913fjbfz874d7-bazel-0.11.1.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=4fdd6f15-b1e0-4839-830c-c4864889d06c)

Attempted: bazel

<details><summary>Partial log (click to expand)</summary><p>

```
//examples/java-native/src/test/java/com/example/myproject:hello
PASSED in 0.4s

Executed 2 out of 2 tests: 2 tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
installing
post-installation fixup
patching script interpreter paths in /nix/store/pmlkwlxlhh1qji7r71iy304m679n1pqk-bazel-0.11.1
checking for references to /tmp in /nix/store/pmlkwlxlhh1qji7r71iy304m679n1pqk-bazel-0.11.1...
/nix/store/pmlkwlxlhh1qji7r71iy304m679n1pqk-bazel-0.11.1
```
</p></details>

LnL7 commented:
@GrahamcOfBorg build bazel

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=903f1f94-cce6-43bb-8b51-abc3a6cb1650)

Attempted: bazel

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/pmlkwlxlhh1qji7r71iy304m679n1pqk-bazel-0.11.1
```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=f4fe4ac9-965b-422e-8704-d22c9eadfa80)

The following builds were skipped because they don't evaluate on aarch64-linux: bazel

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.37608&attempt_id=b4a77d36-6e91-42a7-99d5-d8595bd2069b)

Attempted: bazel

<details><summary>Partial log (click to expand)</summary><p>

```
ld: warning: object file (bazel-out/host/bin/src/main/tools/_objs/linux-sandbox/src/main/tools/dummy-sandbox.o) was built for newer OSX version (10.13) than being linked (10.10)
INFO: From Linking src/main/tools/build-runfiles [for host]:
ld: warning: object file (bazel-out/host/bin/src/main/tools/_objs/build-runfiles/src/main/tools/build-runfiles.o) was built for newer OSX version (10.13) than being linked (10.10)
[460 / 501] Compiling third_party/protobuf/3.4.0/src/google/protobuf/descriptor.cc [for host]; 45s local ... (8 actions running)
INFO: From Linking third_party/protobuf/3.4.0/libprotobuf.a [for host]:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: bazel-out/host/bin/third_party/protobuf/3.4.0/libprotobuf.a(gzip_stream.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: bazel-out/host/bin/third_party/protobuf/3.4.0/libprotobuf.a(error_listener.o) has no symbols
[503 / 543] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_primitive_field.cc [for host]; 5s local ... (8 actions running)
building of '/nix/store/qq2nbfqzx0s02sz57x0kbgv939lnhj64-bazel-0.11.1.drv' timed out after 1800 seconds
[31;1merror:[0m build of '/nix/store/qq2nbfqzx0s02sz57x0kbgv939lnhj64-bazel-0.11.1.drv' failed
```
</p></details>

abbradar commented:
@lukeadams No sweat! I've merged https://github.com/NixOS/nixpkgs/pull/37044 for now.

LnL7 commented:
I was hoping the framework was the only issue, but I get an `xcrun` failure locally. 😕

lukeadams commented:
Closed until I have time to work on this. Not sure when tbh :/

#40424 with `full-paths.patch` patch looks interesting for Darwin
